tests/MongoDB.Driver.Core.Tests/Core/Operations/AsyncCursorTests.cs

```diff
@@ -180,16 +325,113 @@ public void CreateGetMoreCommand_should_return_expected_result_when_maxTime_is_p
         }
 
         [Fact]
-        public void Dispose_should_dispose_channel_source_when_cursor_id_is_zero()
+        public void CreateKillCursorsCommand_should_return_expected_result()
+        {
+            var subject = CreateSubject(cursorId: 1);
+
+            var result = subject.CreateKillCursorsCommand();
+
+            result.Should().Be("{ \"killCursors\" : \"test\", \"cursors\" : [NumberLong(1)] }");
+        }
+
+        [Fact]
+        public void Dispose_should_be_shielded_from_exceptions()
+        {
+            var mockChannelSource = new Mock<IChannelSource>();
+            mockChannelSource
+                .Setup(c => c.GetChannel(It.IsAny<CancellationToken>()))
+                .Throws<Exception>();
+
+            var subject = CreateSubject(cursorId: 1, channelSource: Optional.Create(mockChannelSource.Object));
+
+            Action action = () => subject.Dispose();
+            action.ShouldNotThrow<Exception>();
+        }
+
+        [Fact]
+        public void Dispose_should_dispose_channel_source_when_cursor_was_not_closed_by_exception()
         {
             var mockChannelSource = new Mock<IChannelSource>();
+            mockChannelSource
+                .Setup(c => c.GetChannel(It.IsAny<CancellationToken>()))
+                .Throws<Exception>();
+
             var subject = CreateSubject(cursorId: 1, channelSource: Optional.Create(mockChannelSource.Object));
+            subject.Dispose();
+            mockChannelSource.Verify(c => c.Dispose(), Times.Once);
+        }
+
+        [Fact]
+        public void Dispose_should_dispose_channel_source_when_cursor_id_is_zero()
+        {
+            var mockChannelSource = new Mock<IChannelSource>();
+            var subject = CreateSubject(cursorId: 0, channelSource: Optional.Create(mockChannelSource.Object));
 
             subject.Dispose();
 
             mockChannelSource.Verify(s => s.Dispose(), Times.Once);
         }
 
+        [Fact]
+        public void Dispose_should_dispose_cursor_only_once()
+        {
+            int testCursorId = 1;
+            var mockChannelHandle = new Mock<IChannelHandle>();
+            var mockChannelSource = new Mock<IChannelSource>();
+            SetupChannelMocks(mockChannelSource, mockChannelHandle, false, $"{{ 'ok' : true, 'cursorsNotFound' : [], 'cursorsKilled' : [{testCursorId}] }}");
+
+            var subject = CreateSubject(cursorId: 1, channelSource: Optional.Create(mockChannelSource.Object));
+            subject.Dispose();
+
+            mockChannelSource.Verify(c => c.GetChannel(It.IsAny<CancellationToken>()), Times.Once);
+            subject.Dispose();
+            mockChannelSource.Verify(c => c.GetChannel(It.IsAny<CancellationToken>()), Times.Once);
```

Comment:
I don't understand how these Verify assertions prove that Dispose can be called twice.

Reply:
The idea is to prove that KillCursor command will be called only once.